keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,364 @@
+
+Benefits from this enhancement includes:
+
+* graduating a feature is no longer a breaking change with new flags.
+
+* well-established feature lifecycle convention from Kubernetes.
+
+* a unified place to store all the feature enablement state in the code base for simplicity, and easy to refer to in guarding the implementation code.
+
+* easy to query if a feature is available for an ectd server.
+
+### Goals
+
+* Introduce the `--feature-gates` flag in etcd code base.
+
+* Introduce apis/methods to query the enablement state of a feature.
+
+* Establish lifecycle stages of a feature, and clear criteria of lifecycle progression.
+
+### Non-Goals
+
+* Handle the feature gate enablement on the cluster level. It would be designed in a separate KEP.
+
+* Migrate all existing `--experimental` features to feature gate. 
+  * While feature gate provides a [path to deprecate existing `--experimental` features]((#path-to-migrate-existing-experimental-features)), and we will migrate several `--experimental` during the feature gate implementation as examples, the completion of the KEP does not require migrating all existing `--experimental` features. It can be done as followup tasks.
+
+* Use feature gate as a mechanism to fix bugs.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+It would be easier for developers to add new features without worrying too much about breaking etcd, because the feature would be Alpha first and disabled by default, and new changes would be mostly gated behind the feature gate.
+
+It would be a smoother process to move the feature to the next stage without the need to introduce a potentially breaking change of an `--experimental` flag.
+
+#### Story 2
+
+Users of etcd cluster would be able to find out which features are enabled in the etcd server and decide on how to use them downstream. 
+
+Currently Kubernetes uses [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to check if a feature is supported by etcd, which basically checks the etcd version and compares the version with a hard-coded map of feature availability for different etcd versions.
+
+#### Story 3
+
+In a HA cluster, users should be able to enable/disable a feature for each server without bring down the whole cluster, by restarting the servers with the new feature gate flag one by one. 
+
+#### Story 4
+
+Users of etcd cluster would like to know if they could use a new feature, but it requires all the etcd servers to agree on the feature enablement. Until the feature gate can handle cluster level features (another KEP), we will not expose querying feature enablement of any cluster level features, because we cannot risk possible conflicting settings that might break cluster correctness.
```

Comment:
We should explain why all etcd servers need to agree on feature enablement. Again issue with supporting window. If we sending a proto with a new field will is not detected in grpc (it's backward compatible), this leads to old clusters interpreting the request differently. We don't know whether every cluster member will interpret the proto in the same way.

Reply:
will defer to cluster feature gate kep.

Reply:
SG